patches/v8/0019-Implement-cross-request-context-promise-resolve-hand.patch

```diff
@@ -0,0 +1,302 @@
+From a1890ed9e4b48067770b4d3205a9d554480692e2 Mon Sep 17 00:00:00 2001
+From: James M Snell <jasnell@gmail.com>
+Date: Mon, 16 Sep 2024 09:56:04 -0700
+Subject: [PATCH] Implement cross-request context promise resolve handling
+
+---
+ include/v8-callbacks.h          | 21 +++++++++++++++++++++
+ include/v8-isolate.h            |  2 ++
+ src/api/api.cc                  |  5 +++++
+ src/builtins/promise-resolve.tq |  7 ++++++-
+ src/execution/isolate-inl.h     |  9 +++++++++
+ src/execution/isolate.cc        | 22 ++++++++++++++++++++++
+ src/execution/isolate.h         | 14 +++++++++++++-
+ src/objects/js-promise.h        |  5 +++++
+ src/objects/objects.cc          | 24 ++++++++++++++++++++++++
+ src/runtime/runtime-promise.cc  | 18 ++++++++++++++++++
+ src/runtime/runtime.h           |  3 ++-
+ 11 files changed, 127 insertions(+), 3 deletions(-)
+
+diff --git a/include/v8-callbacks.h b/include/v8-callbacks.h
+index 25f0c8189e5..df7324d4457 100644
+--- a/include/v8-callbacks.h
++++ b/include/v8-callbacks.h
+@@ -471,6 +471,27 @@ using PromiseCrossContextCallback = MaybeLocal<Promise> (*)(Local<Context> conte
+                                                             Local<Promise> promise,
+                                                             Local<Object> tag);
+ 
++/**
++ * PromiseCrossContextResolveCallback is called when resolving or rejecting a
++ * pending promise whose context tag is not strictly equal to the isolate's
++ * current promise context tag. The callback is called with the promise to be
++ * resolved, it's context tag, and a function that when called, causes the
++ * reactions to the resolved promise to be enqueued. The idea is that the
++ * embedder sets this callback in the case it needs to defer the actual
++ * scheduling of the reactions to the given promise to a later time.
++ * Importantly, when this callback is invoked, the state of the promise
++ * should have already been updated. We're simply possibly deferring the
++ * enqueue of the reactions to the promise.
++ */
++using PromiseCrossContextResolveCallback = void (*)(
++    v8::Isolate* isolate,
++    Local<Value> tag,
++    Local<Value> reactions,
```

Comment:
I was hoping this callback could take a `Local<Promise::Resolver>` and just say "please resolve me in the right context". That seems cleaner as it doesn't expose any of the internals of promises and doesn't need an `std::function` since the embedder ends up calling an existing public API.

Maybe this callback should hook in a little bit earlier than it does? I was basically thinking that when you call `resolve()`, the _first thing_ that happens should be to check the tag. If we're in the wrong context, the callback gets invoked without altering the state of the promise at all...

Reply:
That approach would require quite a more invasive approach when it comes to handling things like custom thenables and resolving a promise with a promise (e.g. `resolve({ then(res, rej) { ... } })` and `resolve(promise)` ... since the v8 currently updates the state of the promise before evaluating those. The actual fulfillment or rejection of the promise is deferred until completion of those.

Also, per https://tc39.es/ecma262/#sec-promise-resolve-functions the language spec requires that when `resolve()` or `reject()` are called, the state of the promise is updated *immediately* and synchronously. Deferring the state change would break this spec-defined invariant and may have other side effects.

While we may be able to hide more of the details of the `reactions` and `argument` from the callback, I think the approach taken in this PR is the best both in terms of minimizing the changes to v8 and preserving spec-defined invariants.

Reply:
> That approach would require quite a more invasive approach when it comes to handling things like custom thenables and resolving a promise with a promise (e.g. resolve({ then(res, rej) { ... } }) and resolve(promise) ... since the v8 currently updates the state of the promise before evaluating those. The actual fulfillment or rejection of the promise is deferred until completion of those.

OK I guess that makes sense. Or to put it another way:
* When `resolve(promise2)` is called, we need `promise2` to run in the calling context, not in the resolver's context. So we can't call the callback yet.
* But we do need to mark the original promise resolved at this time. It would be problematic to wait until `promise2` finishes before doing that, otherwise someone else could call `resolve()` during that time with a non-promise value and it would actually preempt `promise2`, incorrectly.
* So, by the time `promise2` resolves, and we need to call the callback, we can't just pass the resolver itself, because it's already marked resolved and so won't work.